src/cv_compute/examples/examplesScript.sml

```diff
@@ -0,0 +1,216 @@
+(* -------------------------------------------------------------------------
+ * This script contains some instructions on how to use the in-kernel
+ * compute primitive, which is exposed by the function cv_computeLib.cv_compute.
+ *
+ * cv_compute is called with a list of *code equations* and a term to reduce.
+ * The term must be of type :cv. A code equation is a theorem of this shape:
+ *
+ *   |- f (x1:cv) ... (xN:cv) = (e:cv)
+ *
+ * The term `e` must be closed under the xi:s. All function symbols in `e` must
+ * have a corresponding code equation in the list passed to compute. The type
+ * :cv is similar to Lisp pairs (but without the nice s-expression syntax). The
+ * following constructors can be used to create values of type :cv:
+ *
+ *   Num  :num -> cv         (* Numbers *)
+ *   Pair :cv -> cv -> cv    (* Pairs   *)
+ *   cv_add :cv -> cv -> cv  (* arithmetic *)
+ *   cv_sub :cv -> cv -> cv
+ *   cv_mul :cv -> cv -> cv
+ *   cv_div :cv -> cv -> cv
+ *   cv_mod :cv -> cv -> cv
+ *
+ *   cv_lt :cv -> cv -> cv   (* less-than *)
+ *   cv_eq :cv -> cv -> cv   (* equality *)
+ *
+ *   cv_ispair :cv -> cv
+ *   cv_fst :cv -> cv
+ *   cv_snd :cv -> cv
+ *
+ * Arithmetic and less-than treats pairs as the number zero. Things of a Boolean
+ * nature treat non-zero numbers as true, and everything else as false. Further-
+ * more, limited let-bindings and function application are supported:
+ *
+ *   let (x:cv) = (y:cv) in (z:cv)        where x is a variable
+ *   f (x1:cv) ... (xN:cv)                where all xi:s are first-order
+ *
+ * See cvScript.sml for more details on the various cv_-operations, and the file
+ * cv_computeLib.sml for a list of axioms that the different symbols must
+ * satisfy.
+ * ------------------------------------------------------------------------- *)
+
+open HolKernel boolLib bossLib cv_computeLib;
+open arithmeticTheory listTheory prim_recTheory cvTheory;
+
+val _ = new_theory "examples";
+
+val op \\ = op THEN;
+val op >- = op THEN1;
+
+(* -------------------------------------------------------------------------
+ * Example 1: factorial
+ *
+ * Below is a definition of the factorial function in the :cv type. There is
+ * also a pure HOL definition, and a proof that the two definitions compute
+ * related values.
+ * ------------------------------------------------------------------------- *)
+
+Definition factc_def:
+  factc n =
+    cv_if (cv_lt n (Num 1))
+          (Num 1)
+          (cv_mul n (factc (cv_sub n (Num 1))))
+Termination
+  WF_REL_TAC `measure cv_size_alt`
+  \\ Cases \\ simp [cv_size_alt_def, CaseEq "bool"]
+End
+
+Definition fact_def:
+  fact n = if n < 1 then 1 else n * fact (n - 1)
+End
```

Comment:
Please use the existing `FACT` constant from arithmeticTheory.

Reply:
Done